Fix #346

Exzender opened the pull request:
1. Enable preview for the first link only in the regular message (when useEmbeds set to 'never'). So if the user wants to see link previews - he should disable embeds in bot config file.
2. Markdown link format works pretty good in regular messages, except the one, when link caption = link url (when text link auto formatted as link by Telegram). So in this solution is to replace full URL in link caption with word 'link' - in that case discord displays link correctly.

Exzender committed:
Make first link (tg->ds) be with thumbnail - remove [<>] clause

Exzender committed:
Fixed: /chatinfo command was not working in TG channels

Exzender commented:
Added one more fix: command with slash does not detected as 'command' on Telegram side - so was not processed.

Exzender committed:
Fixed: relay error (Telegram->Discord) with empty Header (depends on bridge config)

Exzender commented:
One more fix: when sendUsernames set to off message header (from Telegram) becomes empty - bot failed when tried to set empty title for Embed.

Lulalaby commented on src/telegram2discord/setup.ts:
```diff
@@ -152,9 +153,9 @@ export function setup(
-				if (err instanceof telegraf.TimeoutError) {
-					throw err;
-				}
+				// if (err instanceof telegraf.TimeoutError) {
+				// 	throw err;
+				// }
 
```
What's the reason for this

Exzender replied:
bot crashes when it comes to this block

Lulalaby replied:
Well, the comment says it should be like that. So the root problem has to be somewhere else

Exzender replied:
Yes. It crashes on any Error, because no access to telegraf.TimeoutError.
If you think that it's the right way to kill the bot on any Tg error - maybe just leave one row with "throw err" than ?

Lulalaby replied:
oh, let's remove that commented stuff completely then

Exzender committed:
extended error text when relaying to Discord

Exzender committed:
removed not working TimoutError processing

Exzender committed:
Fixed: #363. Only first media remains when message is edited T→D
Fixed: some issues in link formatting T→D

Exzender committed:
update: edit of message with embed

Exzender committed:
update: 'url' tag processing T->D
update: when only 1 link per post - not use Embed in 'auto' mode - to get web preview

Exzender committed:
Telegram's sendMessage extra option 'disable_web_page_preview' declared as config option discord.disableWebPreviewOnTelegram

Lulalaby commented on README.md:
```diff
@@ -102,7 +102,8 @@ As mentioned in the step-by-step installation guide, there is a settings file. H
 	* `discord.crossDeleteOnTelegram`: Whether to also delete the corresponding message on Telegram when one is deleted in Discord
-        * `discord.useEmbeds`: Whether to use embeds for current bridge. Can be `always`, `never`, `auto`. Defaults to `false`
+	* `discord.disableWebPreviewOnTelegram`: Whether to disable links preview when relaying to Telegram
+    * `discord.useEmbeds`: Whether to use embeds for current bridge. Can be `always`, `never`, `auto`. Defaults to `false`
```
the indentation looks off

Exzender replied:
maybe, spaces vs tabs

Lulalaby reviewed (changes requested):
Other than that nice idea

Exzender committed:
spaces in readme

Lulalaby commented:
can you rebase on dev? i've added the channel chat info as hotfix, so we'd have dup code with your pr now.
Sorry for that

Lulalaby committed:
fix: dup code

Lulalaby commented:
nvm, did it myself :)

Lulalaby committed:
fix import order

Lulalaby commented on src/bridgestuff/BridgeSettingsDiscord.ts:
```diff
@@ -4,6 +4,7 @@ export interface BridgeSettingsDiscordProperties {
 	relayJoinMessages: boolean;
 	relayLeaveMessages: boolean;
 	crossDeleteOnTelegram: boolean;
+	disableWebPreviewOnTelegram?: boolean;
```
that doesn't seem to work

Exzender replied:
![Screenshot_20230911_194311](https://github.com/TediCross/TediCross/assets/42044185/6f012878-2e71-4a5a-a9c7-08b4233f0ca2)

It works.
1st Message sent from DS when option is true.
2nd - when it's false.
It does not depend on Discord's hiding method using <> quotes.

Lulalaby replied:
ok weird, maybe I did something wrong, but the `>;` bug is still there.

Exzender replied:
It's because of this replacements:
`export const escapeHTMLSpecialChars = R.compose(
	R.replace(/>/g, "&gt;"),
	R.replace(/</g, "&lt;"),
	R.replace(/&/g, "&amp;")
);`

Lulalaby replied:
Uh.. OK I'm gonna fix that and then we're gonna merge and release.

Exzender replied:
I can't find proper quick-fix for that.

But I'd suggest one small change - remove ONE symbol "`;`"
`R.replace(/>/g, "&gt")`
This works pretty well.

To get fully correct HTML we can later replace all "`$gt`" with "`$gt;`" right before returning the result here
`return htmlCleanup(html, settings);`

Lulalaby replied:
hmm that gets even more weird. i rather set that on a later todo

Lulalaby reviewed (changes requested):
Discords link suppression `<>` gets converted into a weird `>;` result.
